framework/docs/source/how-to-upgrade-to-message-api.rst

```diff
@@ -0,0 +1,384 @@
+
+.. _clientapp_link: ref-api/flwr.client.ClientApp.html
+
+.. |serverapp_link| replace:: ``ServerApp``
+
+.. _serverapp_link: ref-api/flwr.server.ServerApp.html
+
+.. |strategy_link| replace:: ``Strategy``
+
+.. _strategy_link: ref-api/flwr.serverapp.Strategy.html
+
+.. |fedavg_link| replace:: ``FedAvg``
+
+.. _fedavg_link: ref-api/flwr.serverapp.FedAvg.html
+
+.. |message_link| replace:: ``Message``
+
+.. _message_link: ref-api/flwr.common.Message.html
+
+.. |arrayrecord_link| replace:: ``ArrayRecord``
+
+.. _arrayrecord_link: ref-api/flwr.common.ArrayRecord.html
+
+.. |metricrecord_link| replace:: ``MetricRecord``
+
+.. _metricrecord_link: ref-api/flwr.common.MetricRecord.html
+
+.. |configrecord_link| replace:: ``ConfigRecord``
+
+.. _configrecord_link: ref-api/flwr.common.ConfigRecord.html
+
+.. |recorddict_link| replace:: ``RecordDict``
+
+.. _recorddict_link: ref-api/flwr.common.RecordDict.html
+
+Upgrade to Message API
+======================
+
+Welcome to the migration guide for updating your Flower Apps to use Flower's Message
+API! This guide will walk you through the necessary steps to transition from Flower Apps
+based on ``Strategy`` and ``NumPyClient`` to their equivalent using the new Message API.
+This guide is relevant when updating pre-``1.21`` Flower apps to the latest stable
+version.
+
+Let's dive in!
+
+Summary of changes
+------------------
+
+Thousands of Flower Apps have been created using the Strategies and |numpyclient_link|_
+abstractions. With the introduction of the Message API, these apps can now take
+advantage of a more powerful and flexible communication layer with the |message_link|_
+abstraction being its cornerstone. Messages replace the previous `FitIns` and `FitRes`
+data structures (and their equivalent for the other operations) into a single, unified
+and more versatile data structure.
+
+To fully take advantage of the new Message API, you will need to update your app's code
+to use the new message-based communication patterns. This guide will show you how to:
+
+1. Update your |serverapp_link|_ to make use of the new `Message`-based strategies. You
+   won't need to use the `server_fn` anymore. The new strategies make it easier to
+   customize how the different FL rounds are executed, to retrieve results from your
+   run, and more.
+2. Update your |clientapp_link|_ so it operates directly on `Message` objects received
+   from the |serverapp_link|_. You will be able to keep most of the code from your
+   |numpyclient_link|_ implementation but you won't need to create a new class anymore
+   or use the helper `client_fn` function.
+
+.. tip::
+
+    The main payload Message objects carry are of type |recorddict_link|_. You can think
+    of it as a dictionary that can hold other types of records, namely
+    |arrayrecord_link|_, |metricrecord_link|_, and |configrecord_link|_. Please refer to
+    the documentation for each record for all the details on how they can be constructed
```

Comment:
Instead of pointing to the documentation, we should do a mini-intro here as well. Or would this take too much room?

Reply:
done